governance/ApiDeprecation.md

```diff
@@ -0,0 +1,116 @@
+
+### Steps (Guide)
+
+1. Identify the code to be deprecated (something resembling one of the API deprecation examples outlined in the Overview).
+1. Create a new GitHub issue with the name "Deprecate <Function/Type Name>"
+    1. Add the Deprecation label.
+    1. In the Description, add a preliminary Release Note (what do you want to say to the customer about this change).
+    1. Fill in the Release Note Required field to identify which release this deprecation will be going out in.
+1. Add a `O3DE_DEPRECATION_NOTICE(<GHI Number>)` comment right above the API call.
+    1. e.g.
+```c++
+    // O3DE_DEPRECATION_NOTICE(GHI-1234)
+    AZ::Entity* CreateEditorEntity(const char* name) = 0;
+```
+> Aside: To make searching easier, only add API comments to the function/type declarations, not definitions.
+1. Create second GHI to add `AZ_DEPRECATED` to the API.
+    1. Add the Deprecation label.
+    1. Link the issues together.
+    1. _Note: If it isn't possible to use the `AZ_DEPRECATED` macro, create a ticket and make a note it will apply in Release + 2 (so a maximum of only two GHIs need be created in this case)_
+1. _LATER_ (after the first issue is Closed, i.e. a Release has actually happened) create another GitHub issue to remove the code entirely (please see Tracking Information below for details about this).
+    1. Add the Deprecation label.
+    1. Link this issue to the second issue.
+    1. Add the `AZ_DEPRECATED` macro to the code (leaving the `O3DE_DEPRECATION_NOTICE` comment in place to help with searching) and submit (referencing both Git Hub numbers in the CL description).
+1. _MUCH LATER_ (after the second issue is Closed, i.e. the second Release has actually happened)
+    1. Remove the code and create a PR
+    1. Set final issue to be Closed (will be closed by release team)
+    1. You're done!
+
+### Tracking (Informational)
+
+1. When a Release is imminent, sig-documentaion will (rip)grep the repo for all occurrences of `O3DE_DEPRECATION_NOTICE`
+1. They will confirm what stage the issue is in (following the issue chain) and add the info in the Description to the Release Notes.
+1. sig-documentation will then update the issue to Closed
+    1. When this happens, the engineer who began the deprecation will get a notification from GitHub of the change, and at this time they know it is okay to move to the next stage (add `AZ_DEPRECATED` (part 2), delete code (part 3)).
+
+### Find all O3DE_DEPRECATION_NOTICE tags
+
+```bash
+# prerequisites
+# ripgrep - https://github.com/BurntSushi/ripgrep/releases
+# cmder - https://cmder.net/
+ 
+rg O3DE_DEPRECATION_NOTICE\(GHI\-([0-9]+)\) --replace https://github.com/o3de/o3de/issues/$1 --iglob *.{h,hpp,c,cpp,inl,hxx} -o -I | sort | uniq > deprecations.txt
+ 
+# this will generate a list of Git Hub issue numbers to be updated when a Release is imminent.
+```
+
+### Advice
+
+Use of the `AZ_DEPRECATED` macro is now not allowed when first deprecating an API.
```

Comment:
We don't support this yet, but it might be nice for the AZ_DEPRECATED macro to take an engine version as a parameter(s) which would indicate the engine version where the deprecation goes into effect and automatically enable the warning when the engine version define meets or exceeds that version.  This way, engineers would not need to come back and add the the macro later.  They might even get away with not needing a separate O3DE_DEPRECATION_NOTICE macro.  Something like

```
AZ_DEPRECATED(1,0,0,0, void DepFunc(), "DepFunc will automatically be deprecated in engine versions 1.0.0.0 and beyond") { }
```

This gets into the whole need for us to begin updating the engine version build #defines  via Jenkins so we can do stuff like this, of course major, minor and patch versions would only be updated as part of a release.

Reply:
That's a really nice idea, tagging @amznestebanpapp and @lumberyard-employee-dm to get their take on this (might be out of scope for the first pass of these docs but I really like the idea)

Reply:
This was brought up before, about having some kind of time bomb or version bomb when a deprecation notice becomes a hard error, but the problem is that engine version is only set when we actually make an installer build for the deprecation branch.

It would be right before release of a new version that Jenkins would trigger the deprecation error.

Reply:
True 🤔 Hmm would using a date be possible? Is there some `constexpr` trick for getting the timenow and have the warning fire after a certain delay? 😛 sig-release have been discussing 6 month release windows too which makes this even more challenging

Reply:
This is actually not possible and not a good idea

Querying the current world time, local time, user machine time cannot be done as a constant expression.
How would the compiler know exactly when the time it should set should be.
Furthermore time is not absolute. Whose machine time would the initial value be based on. Is that correct according to whatever timing standards apply. Is that time in UTC+0?

The best that can be done is to run a git command to retrieve the earliest tag in which the commit that added the O3DE_DEPRECATION_NOTICE went in.

Later on when a future release is done, git commands can be run to query information about any "releases" ago the change went in.
i.e "Retrieve all commits that added a `O3DE_DEPRECATION_NOTICE` that was added before tag 2111.1"
That would require a combination of use `git show` to retrieve the text of commits and looking only within the added content and not deleted content, followed by using something like [git-name-rev](https://git-scm.com/docs/git-name-rev) for retrieving the tag.
Furthermore the logic to implement is not trivial as it must account for commits that actually removes the deprecation notice line.